src/traced/probes/packages_list/packages_list_data_source.cc

```diff
@@ -16,89 +16,205 @@
-    return;
+  } else {
+    poller_->Start();
+    Tick();
   }
+}
 
-  bool parsed_fully =
-      ParsePackagesListStream(packages_list_packet, fs, package_name_filter_);
-  if (!parsed_fully)
-    packages_list_packet->set_parse_error(true);
+void PackagesListDataSource::Tick() {
+  // Post next task.
+  auto now_ms = base::GetWallTimeMs().count();
+  auto weak_this = weak_factory_.GetWeakPtr();
+  task_runner_->PostDelayedTask(
+      [weak_this] {
+        if (weak_this)
+          weak_this->Tick();
+      },
+      only_write_on_cpu_use_every_ms_ -
+          static_cast<uint32_t>(now_ms % only_write_on_cpu_use_every_ms_));
+
+  WriteIncrementalPacket();
+}
 
-  if (ferror(*fs))
-    packages_list_packet->set_read_error(true);
+void PackagesListDataSource::WriteIncrementalPacket() {
+  std::vector<CpuPerUidTime> cpu_times = poller_->Poll();
 
-  trace_packet->Finalize();
-  writer_->Flush();
+  if (first_time_) {
+    // The first time we poll the poller it will return details of all UIDs,
+    // even ones whose last activity was days ago, so we wait until a subsequent
+    // time for evidence something has actually run.
+    first_time_ = false;
+    return;
+  }
+
+  std::vector<uint32_t> new_uids;
+  for (auto& time : cpu_times) {
+    if (seen_uids_.find(time.uid) == seen_uids_.end()) {
+      seen_uids_.insert(time.uid);
+      new_uids.push_back(time.uid);
+    }
+  }
+
+  if (new_uids.size() > 0) {
+    auto trace_packet = writer_->NewTracePacket();
+    auto* packages_list_packet = trace_packet->set_packages_list();
+    if (packages_parse_error_)
+      packages_list_packet->set_parse_error(true);
+
+    if (packages_read_error_)
+      packages_list_packet->set_read_error(true);
+
+    for (auto uid : new_uids) {
+      auto range = packages_.equal_range(uid);
+      if (range.first == range.second) {
+        if (uid >= kFirstPackageUid) {
+          PERFETTO_ELOG("No package in list for uid %u", uid);
+        }
+        continue;
+      }
+      for (auto it = range.first; it != range.second; ++it) {
+        auto& package = it->second;
+        auto* package_proto = packages_list_packet->add_packages();
+        package_proto->set_name(package.name.c_str(), package.name.size());
+        package_proto->set_uid(package.uid);
+        package_proto->set_debuggable(package.debuggable);
+        package_proto->set_profileable_from_shell(
+            package.profileable_from_shell);
+        package_proto->set_version_code(package.version_code);
+      }
+    }
+  }
 }
 
 void PackagesListDataSource::Flush(FlushRequestID,
                                    std::function<void()> callback) {
-  // Flush is no-op. We flush after the first write.
-  callback();
+  if (only_write_on_cpu_use_every_ms_ == kWriteAllAtStart) {
+    // Flush is no-op. We flush after the first write.
+    callback();
+  } else {
+    writer_->Flush(callback);
+    poller_->Clear();
+  }
+}
+
+void PackagesListDataSource::ClearIncrementalState() {
```

Comment:
I think the only purpose of ClearIncrementalState is to make sure that if all the previous packets emitted were dropped (by ring buffer), the reader would still have a complete picture from this point on.

The only thing I would do here is to clear `seen_uids_`.

Setting first_time_ to true means that we would ignore all the packages seen since the last tick.

Right?